### DISCUSSION TITLE
BPX Roadmap

### DISCUSSION THREAD

**Discussion by rtimms** [FaradayInstitution/BPX#22] [Ideas]
Starting a discussion to share ideas about what should be included and prioritised in the BPX roadmap. We are keen to get input from a broad range of (potential) BPX users, so please reply below with any comments or suggestions. 

Some suggested items to get us started

- 3D thermal (including details of geometry, tab location, etc.)
- Degradation modelling (either physics-based or empirical)

**Comment by rtimms** [comment]
### Sprint planning
We are proposing a sprint to tackle issues #23, #26, and #29. Below is an outline of the issues and how we intend to address them. We would welcome feedback on how important these issues are and how they should be tackled. We would also appreciate it if people are able to share any use cases that motivate tackling these issues. Thanks!

### #23  Blended electrodes

Add support for well-mixed, blended electrodes that contain more than one active material. For example, Graphite and Silicon, as in [this paper](https://www.sciencedirect.com/science/article/pii/S0378775322001604). The proposed solution is to allow a `dict` of electrode parameters where each key corresponds to a numbered electrode material and each item is an `Electrode` class containing parameters for that electrode. 

E.g. 
```python3
    negative_electrode: Electrode = Field(
        alias="Negative electrode",
    )
```
would be replaced by
```python3
    negative_electrode: Dict[str, Electrode] = Field(
        alias="Negative electrode",
    )
```
and you could provide parameters for a composite electrode with two materials as something like
```python3    
{"Negative electrode": {"primary": {electrode parameters here}, "secondary": {electrode parameters here}}}
``` 

### #26 Miscellaneous category 

Allow users to provide parameters that aren't in the standard. The proposed approach is to have an "Additional Parameterisation" field that mimics the "Parameterisation" schema (cell, negative electrode, separator,...) that can be used to provide additional "user beware" parameters. These would all be of `FloatFunctionTable` type (number, permitted function, or tabulated data). It would be up to the end user to decide how to parse these into the correct format for their simulation tool of choice. 


### #29 Validation based on model

At present, the schema forces the user to provide all the parameters required by the DFN in order for the BPX to be valid. We purpose that the validation should depend on the model type so that if, for example, the user specifies the BPX is for SPM they are allowed to provide a subset of parameters. Initially, BPX support will be for SPM, SPMe and DFN, as defined in this [continuum model review paper](https://iopscience.iop.org/article/10.1088/2516-1083/ac7d31?gclid=CjwKCAjw3K2XBhAzEiwAmmgrAhIelwpWeWtXZoW-vkPQbLoAGtxsqS1AUHas3dCY6cgyY64ZKUXC9xoC1ZUQAvD_BwE). The parameters required for each model are shown in Table 3 of the linked paper.

**Comment by rtimms** [comment]
We also discussed adding simulation metadata (see https://github.com/pybamm-team/BPX/issues/28) to the validation section. This would allow BPX to contain information about how a model was set up and solved in a particular environment in order to valid it against experimental data. We would welcome feedback on how best to include such information and on what information would be necessary to provide confidence in the models and parameters.

**Comment by tommaull** [comment]
Hi all, Peter asked me to chime in here, and I figured if I didn't stir up a little trouble it would seem rude. So here goes:

As industry quickly moves to LFP (and higher silicon anodes) hysteresis effects are becoming more important to understand. I believe that Pybamm already has hysteresis functionality, essentially switching between OCVs. This is a (rough) start but still helpful. However, if you implement the OCV switching functionality, the rest of the parameters begin to lose meaning, as anything that is a function of overpotential changes dramatically. IE parameterization has to include hys or not from the start, and I'd argue that it should. I think we'll see a lot more emphasis on hys in the near future, and what was a simple OCV switch will probably become a lot more nuanced, but as a placeholder I think it has value. Further parameters such as rate of decay might also be useful, either as a constant, or as an SOC dependent function.

**Reply by jamiemfoster** [reply]
Hi Tom,
I'm with you on hysteresis becoming a priority to get right, and I'm strongly of the opinion that OCV switching is a band aid. I'd probably go a bit further to say that it might be a damaging band aid in the long term because it obfuscates the underlying reason that hysteresis occurs in the first place. Would you be interested in having a side bar conversation with myself, Rob, Edmund and anyone else interested on this topic?
Cheers!

**Reply by tommaull** [reply]
Hi Jamie, yes, absolutely keen to have that discussion and especially how it changes parameterization. Will shoot you an email from my work account. Thanks!

**Reply by tommaull** [reply]
Hi Jamie and Rob, really appreciate the chat. I think we all agreed that hysteresis is important to include in PbMs, and will have some significant knock on effects in the parameterisation, so parameters need to be derived/optimised with hysteresis in mind.

**Reply by tommaull** [reply]
Hi everyone, wanted to follow up on this. Based on some LFP work, which hopefully we can make a bit more public in the future, we're more convinced than ever that decay rates need to be included as a function.

**Comment by ikorotkin** [comment]
> We also discussed adding simulation metadata (see #28) to the validation section. This would allow BPX to contain information about how a model was set up and solved in a particular environment in order to valid it against experimental data. We would welcome feedback on how best to include such information and on what information would be necessary to provide confidence in the models and parameters.

To begin addressing this matter, shall we add a new optional field `Meta` (or `Metadata`, `Simulation metadata`?) to each experiment in `Validation`?
`Meta` would contain a few (optional) fields like `Method` (string), `Negative electrode` -> `Number of grid points`, `Number of grid points in particles`, etc.
`Meta` would allow the user to add extra user-defined fields.

Example with pre-defined but optional fields in `Meta`:
```json
        "Validation": {
            "Experiment 1": {
                "Meta": {
                    "Method": "Finite Volumes",
                    "URL": "https://link-to-the-sim-results",
                    "Solver": "MATLAB ode15s",
                    "Maximum time step [s]": 10.0,
                    "Absolute tolerance": 1e-6,
                    "Relative tolerance": 1e-6,
                    "Negative electrode": {
                        "Number of grid points": 16,
                        "Number of grid points in particles": 16,
                    },
                    "Positive electrode": {
                        "Number of grid points": 16,
                        "Number of grid points in particles": 16,
                    },
                    "Separator": {
                        "Number of grid points": 16,
                    },
                },
                "Time [s]": [0, 1000, 2000],
                "Current [A]": [-0.625, -0.625, -0.625],
                "Voltage [V]": [4.19367569, 4.1677888, 4.14976386],
                "Temperature [K]": [298.15, 298.15, 298.15],
            },
        }
```
